src/ledger/GeneralizedLedgerEntry.cpp

```diff
@@ -0,0 +1,609 @@
+{
+    return lhs.sponsoredID == rhs.sponsoredID &&
+           lhs.sponsoringID == rhs.sponsoringID;
+}
+
+bool
+operator!=(SponsorshipEntry const& lhs, SponsorshipEntry const& rhs)
+{
+    return !(lhs == rhs);
+}
+
+bool
+operator==(SponsorshipCounterKey const& lhs, SponsorshipCounterKey const& rhs)
+{
+    return lhs.sponsoringID == rhs.sponsoringID;
+}
+
+bool
+operator!=(SponsorshipCounterKey const& lhs, SponsorshipCounterKey const& rhs)
+{
+    return !(lhs == rhs);
+}
+
+bool
+operator==(SponsorshipCounterEntry const& lhs,
+           SponsorshipCounterEntry const& rhs)
+{
+    return lhs.sponsoringID == rhs.sponsoringID &&
+           lhs.numSponsoring == rhs.numSponsoring;
+}
+
+bool
+operator!=(SponsorshipCounterEntry const& lhs,
+           SponsorshipCounterEntry const& rhs)
+{
+    return !(lhs == rhs);
+}
+
+// GeneralizedLedgerKey -------------------------------------------------------
+GeneralizedLedgerKey::GeneralizedLedgerKey()
+    : GeneralizedLedgerKey(GeneralizedLedgerEntryType::LEDGER_ENTRY)
+{
+}
+
+GeneralizedLedgerKey::GeneralizedLedgerKey(GeneralizedLedgerEntryType t)
+    : mType(t)
+{
+    construct();
+}
+
+GeneralizedLedgerKey::GeneralizedLedgerKey(LedgerKey const& lk)
+    : GeneralizedLedgerKey(GeneralizedLedgerEntryType::LEDGER_ENTRY)
+{
+    ledgerKey() = lk;
+}
+
+GeneralizedLedgerKey::GeneralizedLedgerKey(SponsorshipKey const& sk)
+    : GeneralizedLedgerKey(GeneralizedLedgerEntryType::SPONSORSHIP)
+{
+    sponsorshipKey() = sk;
+}
+
+GeneralizedLedgerKey::GeneralizedLedgerKey(SponsorshipCounterKey const& sck)
+    : GeneralizedLedgerKey(GeneralizedLedgerEntryType::SPONSORSHIP_COUNTER)
+{
+    sponsorshipCounterKey() = sck;
+}
+
+GeneralizedLedgerKey::GeneralizedLedgerKey(GeneralizedLedgerKey const& glk)
+    : GeneralizedLedgerKey(glk.type())
+{
+    assign(glk);
+}
+
+GeneralizedLedgerKey::GeneralizedLedgerKey(GeneralizedLedgerKey&& glk)
+    : GeneralizedLedgerKey(glk.type())
+{
+    assign(std::move(glk));
+}
+
+GeneralizedLedgerKey&
+GeneralizedLedgerKey::operator=(GeneralizedLedgerKey const& glk)
+{
+    type(glk.type());
+    assign(glk);
+    return *this;
+}
+
+GeneralizedLedgerKey&
+GeneralizedLedgerKey::operator=(GeneralizedLedgerKey&& glk)
+{
+    if (this == &glk)
+    {
+        return *this;
+    }
+
+    type(glk.type());
+    assign(std::move(glk));
+    return *this;
+}
+
+GeneralizedLedgerKey::~GeneralizedLedgerKey()
+{
+    destruct();
+}
+
+void
+GeneralizedLedgerKey::assign(GeneralizedLedgerKey const& glk)
+{
+    assert(glk.type() == mType);
+    switch (mType)
+    {
+    case GeneralizedLedgerEntryType::LEDGER_ENTRY:
+        ledgerKey() = glk.ledgerKey();
+        break;
+    case GeneralizedLedgerEntryType::SPONSORSHIP:
+        sponsorshipKey() = glk.sponsorshipKey();
+        break;
+    case GeneralizedLedgerEntryType::SPONSORSHIP_COUNTER:
+        sponsorshipCounterKey() = glk.sponsorshipCounterKey();
+        break;
+    default:
+        abort();
+    }
+}
```

Comment:
It would be possible to define something like:

```
template <typename Ret, typename LedgerEntryFunc, typename SponsorshipFunc,
          typename SponsorshipCounterFunc>
Ret
GeneralizedLedgerKey::match(LedgerEntryFunc leFunc, SponsorshipFunc spFunc,
                            SponsorshipCounterFunc scFunc)
{
    switch (mType)
    {
    case GeneralizedLedgerEntryType::LEDGER_ENTRY:
        return leFunc();
        break;
    case GeneralizedLedgerEntryType::SPONSORSHIP:
        return spFunc();
        break;
    case GeneralizedLedgerEntryType::SPONSORSHIP_COUNTER:
        return scFunc();
        break;
    default:
        abort();
    }
}
```

And then the above would become:

```
void
GeneralizedLedgerKey::assign(GeneralizedLedgerKey const& glk)
{
    assert(glk.type() == mType);
    match<void>(
        [&]() { ledgerKey() = glk.ledgerKey(); },
        [&]() { sponsorshipKey() = glk.sponsorshipKey(); },
        [&]() { sponsorshipCounterKey() = glk.sponsorshipCounterKey(); });
}
```

...with a similar shortening to several following methods.  (The same could be done with `GeneralizedLedgerEntry`.)

I like that it gets the boilerplate `switch`/`case`/`break`/`default` skeleton out of those methods, and is probably shorter overall, and makes the compiler tell you if you don't add a case to one of them when adding an entry type in the future.  But I could imagine someone hating the style, too.  ;)  Just a thought.